### PR TITLE
Add warnings if skill cost exceeds avilable resource.

### LINKED ISSUE #4998: Add a warning when auras are disabled and disable if you cannot cast them.
On second though it would probably be better to drop the automatic disabling add only provide warnings.

### Check for duplicates

- [X] I've checked for duplicate issues by using the search function of the [issue tracker](https://github.com/PathOfBuildingCommunity/PathOfBuilding/issues)

### Is your feature request related to a problem?

Currently pob disables aura if there reservation exceeds maximum resource. This should be extended to cast cost and provide a warning as it can be annoying not understanding why level 20 clarity isn't working at a level 1 default pob config.

### Describe the solution you'd like

Add a warning when aura is disabled and disable auras that cannot be cast. e.g. divine blessing. This could be extended to skills as well.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Paliak**
Fixes #4998 .

### Description of the problem being solved:
Currently there is no warning that a skill uses more life/mana/Es/Rage than is available. This pr also stops auras from not applying their effects even though reservation is applied even when above possible reservation.

Note that this pr makes triggered skills get cached. This may or may not cause issues with some calculations but i was not able to find any issues.
### Steps taken to verify a working solution:
- Check skill setups
- Check divine blessing interaction
- Check auras while having two essence worm rings applied the supreme ego allocated.

### Before screenshot:
obraz

### After screenshot:
obraz

### CODE REVIEW

File: src/Modules/Build.lua [hunk 1]
```diff
@@ -318,8 +318,8 @@ function buildMode:Init(dbFileName, buildName, buildXML, convertBuild)
...
 		{ stat = "BrandAttachmentRange", label = "Attachment Range", fmt = "d", flag = "brand" },
 		{ stat = "BrandTicks", label = "Activations per Brand", fmt = "d", flag = "brand" },
-		{ stat = "ManaCost", label = "Mana Cost", fmt = "d", color = colorCodes.MANA, compPercent = true, lowerIsBetter = true, condFunc = function(v,o) return o.ManaHasCost end },
-		{ stat = "LifeCost", label = "Life Cost", fmt = "d", color = colorCodes.LIFE, compPercent = true, lowerIsBetter = true, condFunc = function(v,o) return o.LifeHasCost end },
+		{ stat = "ManaCost", label = "Mana Cost", fmt = "d", color = colorCodes.MANA, pool = "Mana", compPercent = true, lowerIsBetter = true, condFunc = function(v,o) return o.ManaHasCost end, warnFunc = function(v,o) return (o.ManaCostWarning or (v > o.ManaUnreserved)) and "You do not have enough Mana to use a Selected Skill" end},
+		{ stat = "LifeCost", label = "Life Cost", fmt = "d", color = colorCodes.LIFE, pool = "Life", compPercent = true, lowerIsBetter = true, condFunc = function(v,o) return o.LifeHasCost end, warnFunc = function(v,o) return (o.LifeCostWarning or (v > o.LifeUnreserved)) and "You do not have enough Mana to use a Selected Skill" end },
```
Comment (QuickStick123) [inline src/Modules/Build.lua:322]:
```suggestion
		{ stat = "LifeCost", label = "Life Cost", fmt = "d", color = colorCodes.LIFE, pool = "Life", compPercent = true, lowerIsBetter = true, condFunc = function(v,o) return o.LifeHasCost end, warnFunc = function(v,o) return (o.LifeCostWarning or (v > o.LifeUnreserved)) and "You do not have enough Life to use a Selected Skill" end },
```
Consider adding support for rage and energy shield cost as well. These warning might also be inaccurate with eldrich battery but I am not sure the best way to address this.

Comment (Paliak) [reply src/Modules/Build.lua]: Fixed by latest commit. Warnings for Energy shield and rage have been implemented.

Comment (QuickStick123) [reply src/Modules/Build.lua]: You forgot to commit the energy shield / rage implementation?

Comment (Paliak) [reply src/Modules/Build.lua]: Yup... Really need to pay more attention to what i'm doing.

File: src/Modules/Build.lua [hunk 2]
```diff
@@ -318,10 +318,10 @@ function buildMode:Init(dbFileName, buildName, buildXML, convertBuild)
...
-		{ stat = "RageCost", label = "Rage Cost", fmt = "d", color = colorCodes.RAGE, compPercent = true, lowerIsBetter = true, condFunc = function(v,o) return o.RageHasCost end },
+		{ stat = "ManaCost", label = "Mana Cost", fmt = "d", color = colorCodes.MANA, pool = "Mana", compPercent = true, lowerIsBetter = true, condFunc = function(v,o) return o.ManaHasCost end, warnFunc = function(v,o) return (o.EnergyShieldProtectsMana and (v > (o.ManaUnreserved + o.EnergyShield)) or (v > o.ManaUnreserved)) and "You do not have enough Mana to use a Selected Skill" end },
+		{ stat = "LifeCost", label = "Life Cost", fmt = "d", color = colorCodes.LIFE, pool = "Life", compPercent = true, lowerIsBetter = true, condFunc = function(v,o) return o.LifeHasCost end, warnFunc = function(v,o) return (v > o.LifeUnreserved) and "You do not have enough Life to use a Selected Skill" end },
+		{ stat = "ESCost", label = "Energy Shield Cost", fmt = "d", color = colorCodes.ES, pool = "EnergyShield", compPercent = true, lowerIsBetter = true, condFunc = function(v,o) return o.ESHasCost end, warnFunc = function(v,o) return (v > o.EnergyShield) and "You do not have enough EnergyShield to use a Selected Skill" end },
```
Comment (QuickStick123) [inline src/Modules/Build.lua:323]:
```suggestion
		{ stat = "ESCost", label = "Energy Shield Cost", fmt = "d", color = colorCodes.ES, pool = "EnergyShield", compPercent = true, lowerIsBetter = true, condFunc = function(v,o) return o.ESHasCost end, warnFunc = function(v,o) return (v > o.EnergyShield) and "You do not have enough Energy Shield to use a Selected Skill" end },
```

File: src/Modules/Build.lua [hunk 3]
```diff
@@ -318,10 +318,10 @@ function buildMode:Init(dbFileName, buildName, buildXML, convertBuild)
...
 		{ stat = "BrandTicks", label = "Activations per Brand", fmt = "d", flag = "brand" },
-		{ stat = "ManaCost", label = "Mana Cost", fmt = "d", color = colorCodes.MANA, compPercent = true, lowerIsBetter = true, condFunc = function(v,o) return o.ManaHasCost end },
-		{ stat = "LifeCost", label = "Life Cost", fmt = "d", color = colorCodes.LIFE, compPercent = true, lowerIsBetter = true, condFunc = function(v,o) return o.LifeHasCost end },
-		{ stat = "ESCost", label = "Energy Shield Cost", fmt = "d", color = colorCodes.ES, compPercent = true, lowerIsBetter = true, condFunc = function(v,o) return o.ESHasCost end },
-		{ stat = "RageCost", label = "Rage Cost", fmt = "d", color = colorCodes.RAGE, compPercent = true, lowerIsBetter = true, condFunc = function(v,o) return o.RageHasCost end },
+		{ stat = "ManaCost", label = "Mana Cost", fmt = "d", color = colorCodes.MANA, pool = "Mana", compPercent = true, lowerIsBetter = true, condFunc = function(v,o) return o.ManaHasCost end, warnFunc = function(v,o) return (o.EnergyShieldProtectsMana and (v > (o.ManaUnreserved + o.EnergyShield)) or (v > o.ManaUnreserved)) and "You do not have enough Mana to use a Selected Skill" end },
```
Comment (QuickStick123) [inline src/Modules/Build.lua:321]:
```suggestion
		{ stat = "ManaCost", label = "Mana Cost", fmt = "d", color = colorCodes.MANA, pool = "Mana", compPercent = true, lowerIsBetter = true, condFunc = function(v,o) return o.ManaHasCost end, warnFunc = function(v,o) return (not o.EnergyShieldProtectsMana and (v > o.ManaUnreserved) or (v > o.ManaUnreserved + o.EnergyShield)) and "You do not have enough "..(o.EnergyShieldProtectsMana and "Energy Shield and Mana" or "Mana").." to use a Selected Skill" end },
```
The issue here is that the following condition could be true if the left wasn't this switches the order to ensure we don't try the regular check on EB. Also improve tooltip.

File: src/Modules/Build.lua [hunk 4]
```diff
@@ -1261,13 +1261,16 @@ function buildMode:RefreshSkillSelectControls(controls, mainGroup, suffix)
...
 
-function buildMode:FormatStat(statData, statVal, overCapStatVal)
+function buildMode:FormatStat(statData, statVal, overCapStatVal, pool)
 	if type(statVal) == "table" then return "" end
 	local val = statVal * ((statData.pc or statData.mod) and 100 or 1) - (statData.mod and 100 or 0)
 	local color = (statVal >= 0 and "^7" or statData.chaosInoc and "^8" or colorCodes.NEGATIVE)
 	if statData.label == "Unreserved Life" and statVal == 0 then
 		color = colorCodes.NEGATIVE
 	end
+	if statData.label and statData.label:match("Cost$") and statVal and pool and statVal > pool then
```
Comment (QuickStick123) [inline src/Modules/Build.lua:1271]: This incorrectly marks EB as a red cost.
image

Comment (Paliak) [reply src/Modules/Build.lua]: Should be fixed by https://github.com/PathOfBuildingCommunity/PathOfBuilding/pull/5019/commits/b6705ebfa0964af51ce5958f1735d5812389786b

**Comment by QuickStick123** [comment]
Text warning for life/energy shield now doesn't seem to appear. Everything else seems fine.

**Comment by Paliak** [comment]
@QuickStick123 I assume other costs would be in the top level of the cache table too but that was not the case. https://github.com/PathOfBuildingCommunity/PathOfBuilding/pull/5019/commits/946c47e48f2a8fcc90a5ff28876ff63246109f9e should fix it. Thanks for the help.

**Review by QuickStick123 — APPROVED** [review]
LGTM the only remain issue is output.Rage doesn't exist in current dev.
This can be added here as well or can just wait until https://github.com/PathOfBuildingCommunity/PathOfBuilding/pull/5011 is merged. Up to you.

**Comment by Paliak** [comment]
@QuickStick123 I made it default to 0 if output.Rage is nil so i think it's fine if it stays separate.

**Comment by LocalIdentity** [comment]
This will need a couple fixes for conflicts generated by https://github.com/PathOfBuildingCommunity/PathOfBuilding/pull/5199 being merged

**Review by Wires77 — APPROVED** [review]
LGTM

**Pull request merged**